File: react-native-pytorch-core/android/src/main/java/org/pytorch/rn/core/image/Image.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.rn.core.image;

import android.content.Context;
import android.graphics.Bitmap;
import androidx.annotation.Nullable;
import androidx.camera.core.ImageProxy;
import org.pytorch.rn.core.canvas.ImageData;

public class Image implements IImage {
  private final IImage mImage;

  public Image(Bitmap bitmap) {
    mImage = new BitmapImage(bitmap);
  }

  public Image(ImageProxy imageProxy, Context context) {
    mImage = new ImageProxyImage(imageProxy, context);
  }

  public Image(ImageData imageData, float pixelDensity) {
    mImage = new ImageDataImage(imageData, pixelDensity);
  }

  @Override
  public float getPixelDensity() {
    return mImage.getPixelDensity();
  }

  @Override
  public int getWidth() {
    return mImage.getWidth();
  }

  @Override
  public int getHeight() {
    return mImage.getHeight();
  }

  @Override
  public float getNaturalWidth() {
    return mImage.getNaturalWidth();
  }

  @Override
  public float getNaturalHeight() {
    return mImage.getNaturalHeight();
  }

  @Override
  public IImage scale(float sx, float sy) {
    return mImage.scale(sx, sy);
  }

  @Override
  public Bitmap getBitmap() {
    return mImage.getBitmap();
  }

  @Override
  public @Nullable android.media.Image getImage() {
    return mImage.getImage();
  }

  @Override
  public int getImageRotationDegrees() {
    return mImage.getImageRotationDegrees();
  }

  @Override
  public void close() throws Exception {
    mImage.close();
  }
}

File: react-native-pytorch-core/ios/Image/Image.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

enum ImageError: Error {
    case scale
}

public class Image: IImage {

    private var image: IImage
    private var closed: Bool = false

    init(image: CGImage) {
        self.image = BitmapImage(image: image)
    }

    init(image: CIImage) {
        self.image = BitmapImage(image: image)
    }

    init(imageData: ImageData, pixelDensity: CGFloat) {
        self.image = ImageDataImage(imageData: imageData, pixelDensity: pixelDensity)
    }

    public func getPixelDensity() -> CGFloat {
        return self.image.getPixelDensity()
    }

    public func getWidth() -> CGFloat {
        return self.image.getWidth()
    }

    public func getHeight() -> CGFloat {
        return self.image.getHeight()
    }

    public func getNaturalWidth() -> Int {
        // Note: on Android, returns the width in device pixels so that it can be properly redrawn back to the canvas
        // Since iOS always uses points (density independent), this will return the same as getWidth so that it can
        // be properly redrawn back to the canvas
        return self.image.getNaturalWidth()
    }

    public func getNaturalHeight() -> Int {
        // Note: on Android, returns the height in device pixels so that it can be properly redrawn back to the canvas
        // Since iOS always uses points (density independent), this will return the same as getHeight so that it can
        // be properly redrawn back to the canvas
        return self.image.getNaturalHeight()
    }

    public func scale(sx: CGFloat, sy: CGFloat) throws -> IImage {
        return try self.image.scale(sx: sx, sy: sy)
    }

    public func getBitmap() -> CGImage? {
        return self.image.getBitmap()
    }

    public func close() throws {
        self.closed = true
    }

    public func isClosed() -> Bool {
        return self.closed
    }
}
